### PR TITLE
[RPS-447] Use S3 mulipart upload

### PR DESCRIPTION
**Pull request by wojtek-oledzki**
The externalstorage:resource upload dialog is now uploading files in 5MB
parts, which allows us to upload up to 50GB files.

### CODE REVIEW

File: cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java [hunk 1]
```diff
@@ -54,9 +90,44 @@ public S3ObjectMetadata uploadFile(InputStream fileStream, String fileName, Stri
...
+                try (ByteArrayInputStream partStream = new ByteArrayInputStream(buffer, 0, lastBytesReadCount)) {
+                    UploadPartRequest req = new UploadPartRequest()
+                        .withBucketName(bucketName)
+                        .withKey(objectKey)
+                        .withUploadId(uploadId)
+                        .withInputStream(partStream)
+                        .withPartNumber(currentPartNumber)
+                        .withPartSize(lastBytesReadCount);
+
+                    partETags.add(s3.uploadPart(req).getPartETag());
```
Comment (ziemsky) [inline cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java]: It appears that the unit test for `S3Connector` does not verify that the upload of individual chunks actually takes place (the `s3.upload` call).

File: cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java [hunk 2]
```diff
@@ -45,7 +67,21 @@ public S3ObjectMetadata uploadFile(InputStream fileStream, String fileName, Stri
...
+        InitiateMultipartUploadResult initResult = s3.initiateMultipartUpload(new InitiateMultipartUploadRequest(bucketName, objectKey, metadata));
+        initResult.getUploadId();
+
+        // loop parts
+        List<PartETag> partETags = new ArrayList<>();
+        try {
+            partETags = uploadParts(fileStream, bucketName, objectKey, initResult.getUploadId());
+        } catch (IOException ex) {
+            log.error("Error while uploading file", ex);
+            s3.abortMultipartUpload(new AbortMultipartUploadRequest(bucketName, objectKey, initResult.getUploadId()));
```
Comment (ziemsky) [inline cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java]: The `S3Connector` unit test needs to verify that the request gets aborted on failure.

File: cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java [hunk 3]
```diff
@@ -45,7 +67,21 @@ public S3ObjectMetadata uploadFile(InputStream fileStream, String fileName, Stri
...
+        }
+
+        // finalise multipart upload
+        s3.completeMultipartUpload(new CompleteMultipartUploadRequest(bucketName, objectKey, initResult.getUploadId(), partETags));
```
Comment (ziemsky) [inline cms/src/main/java/uk/nhs/digital/externalstorage/s3/S3ConnectorImpl.java]: The unit test doesn't seem to verify that the 'complete request' call takes place.

**Comment by ziemsky** [comment]
I've expanded the unit test, @wojtek-oledzki, and slightly modified exception handling - when 'abort request' is triggered, no attempt of 'complete request' nor 'get metadata' is made, but exception is thrown instead.

Would you mind having a look if you agree?

**Comment by ziemsky** [comment]
Now rebased onto latest `master`.

**Comment by wojtek-oledzki** [comment]
Pure Awesomeness!

Pure Awesomeness

**Pull request merged**